fix(EIDOMNI-1159): update cookbook documentation (#471)
swiyu-admin-ch/swiyu-issuer: docs/deferred.md

end

# Get credential
WALLET->>+ISS: Get openid metadata
ISS-->>-WALLET:

WALLET->>+ISS: Get issuer metadata
ISS-->>-WALLET:

WALLET->>+ISS: Get oauth token
ISS-->>-WALLET: Oauth token

WALLET->>+ISS: Redeem offer
ISS->>+DB: Get offer data and status list INFO
DB-->-ISS:
ISS->>+DB: Set STATUS = Deferred
DB-->-ISS:
ISS-->>-WALLET : Transaction id

loop get status
BUSINESS->>+ISS: Get status
ISS-->>-BUSINESS: Status

alt STATUS is Deferred
BUSINESS->>BUSINESS: Some additional process
alt offer data is already set
BUSINESS->>+ISS: Set status READY
else
BUSINESS->>+ISS : Set offer data (STATUS is set to READY)
end
ISS->>DB: Store offer
ISS-->>-BUSINESS:
end
end

loop Get deferred credential
alt STATUS is not READY
WALLET->>+ISS: Get credential from deferred_credential
ISS->>+DB: Get offer data and status list INF
DB-->-ISS:
ISS-->>-WALLET: issuance_pending
else
WALLET->>+ISS: Get credential from deferred_credential
ISS->>+DB: Get offer data and status list INFO
DB-->-ISS:
ISS-->>-WALLET: VC
end
end

loop STATUS is ISSUED
BUSINESS->>+ISS: Get status
ISS->>+DB: Remove offer data
ISS-->>-BUSINESS: Status
end
```

### Create status list entry

Actor: Business Issuer

> [!NOTE]  

> > In order to create an offer first you have to initialize a status list. Please store the `statusRegistryUrl` as it is

Reviewer: ```suggestion
> In order to create an offer you first have to initialize a status list. Please store the `statusRegistryUrl` as it is
```

* `maxLength`: The maximum number of entries in the status list. This is a hard limit and cannot be changed later.
* `config`: Configuration for the status list. The `bits` parameter defines the number of bits used for the status list
  entry.
    - `bits: 2` means that each entry in the status list can have 2 different states (issued, revoked, suspended).

```bash
curl -X 'POST' \
  'http://localhost:8080/management/api/status-list' \
  -H 'accept: */*' \
  -H 'Content-Type: application/json' \
  -d '{
  "type": "TOKEN_STATUS_LIST",
  "maxLength": 100000,
  "config": {
    "bits": 2
  }
}'
```

With the response:

```json
{
    "id": "36f98a79-4be7-4978-bc46-071d8e40343a",
    "statusRegistryUrl": "your new $STATUS_REGISTRY_URL",
    "type": "TOKEN_STATUS_LIST",
    "maxListEntries": 100000,
    "remainingListEntries": 100000,
    "version": "1.0",
    "config": {
        "bits": 2
    }
}
```

### Create credential offer

> [!NOTE]  
> In order to create a deferred credential offer, you need to set `"deferred": true` in the `credential_metadata`.
> In contrast to the normal credential offer, the `credential_subject_data` is not required in the request body, as it

> > can be set later, as the content might not be known beforehand (`"credential_subject_data": {}`).

Reviewer: ```suggestion
> can be set later, because the content might not be known beforehand (`"credential_subject_data": {}`).
```

```bash
curl -X 'POST' \
  'http://localhost:8080/management/api/credentials' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/json' \
  -d '{
  "metadata_credential_supported_id": [
    "university_example_sd_jwt"
  ],
  "credential_subject_data": {
    "type": "Bachelor of Science",
    "name":"Data Science",
    "average_grade":"5.33"
  },
  "credential_metadata": {
    "deferred": true
  },
  "offer_validity_seconds": 86400,
  "deferred_offer_validity_seconds": 86400,
  "credential_valid_until": "2030-01-01T19:23:24Z",
  "credential_valid_from": "2025-01-01T18:23:24Z",
  "status_lists": [
    "$STATUS_REGISTRY_URL"
  ]
}'
```

> The response will look like this:

Reviewer: ```suggestion
The response looks like this:
```

```json
{
    "management_id": "6444cd65-f550-4934-af85-1bd295cea1ba",
    "offer_deeplink": "swiyu://?credential_offer=%7B%22grants%22%3A%7B%22urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Apre-authorized_code%22%3A%7B%22pre-authorized_code%22%3A%2206b5424f-8baf-4eaa-a770-7fa8ab7a85a7%22%7D%7D%2C%22version%22%3A%221.0%22%2C%22credential_issuer%22%3A%22http%3A%2F%2Flocalhost%3A8080%2F%22%2C%22credential_configuration_ids%22%3A%5B%22university_example_sd_jwt%22%5D%7D"
}
```

> [!NOTE]  
> The `offer_deeplink` can then be passed to the wallet application as QR-Code, which will continue with the credential
> process.
> Please store the `management_id` as it is needed in later steps and will be referenced as `$MANAGEMENT_ID`.
> Also store the `offer_deeplink` as it is needed to create an access token in the next step.

### Get credential offer status with the $MANAGEMENT_ID

```bash
curl -X 'GET' 'http://localhost:8080/management/api/credentials/$MANAGEMENT_ID/status' -H 'accept: application/json'
```

> The response will look like this:

Reviewer: ```suggestion
The response looks like this:
```

```json
{
    "status": "OFFERED"
}
```

### Get Metadata for the credential offer

Actor: Wallet

The wallet will fetch the OpenID Connect metadata from the Issuer to know how to get an access token.

```bash
curl -X 'GET' 'http://localhost:8080/.well-known/openid-configuration' -H 'accept: application/json'
```

With response:

```json
{
    "issuer": "http://localhost:8080",
    "token_endpoint": "http://localhost:8080/oid4vci/api/token"
}
```

> The wallet will also fetch the Issuer metadata to know which credential configurations are supported and witch

Reviewer: 🧙‍♀️
```suggestion
The wallet will also fetch the Issuer metadata to know which credential configurations are supported and which
```

}
```

### Get access token for the credential offer

Actor: Wallet

The wallet decodes the `offer_deeplink` and extracts the `pre-authorized_code` value. The wallet then uses this code to
get a Bearer Token. The request body must be URL-encoded and the `grant_type` must be set
to `urn:ietf:params:oauth:grant-type:pre-authorized_code`.

```bash
curl -X 'POST' \
  'http://localhost:8080/oid4vci/api/token' \
  -H 'accept: application/json' \
  -H 'Content-Type: application/x-www-form-urlencoded' \
  -d 'grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Apre-authorized_code&pre-authorized_code=THIS_IS_THE_PRE_AUTHORIZED_CODE_FROM_THE_OFFER_DEEPLINK'
```

With response:

```json
{
    "access_token": "your $ACCESS_TOKEN",
    "expires_in": 600,
    "token_type": "BEARER"
}
```

> [!NOTE]
> Please store the `access_token` as it is needed in later steps and will be referenced as `$ACCESS_TOKEN`.

### Get Nonce

Actor: Wallet

```bash
curl -X 'POST' 'http://localhost:8080/oid4vci/api/nonce' -H 'accept: application/json'
```

With response:

```json
{
    "c_nonce": "c9c737f3-f37b-4331-a23c-313dc821fac5::2025-08-07T16:30:19.021045078Z"
}
```

> [!NOTE]

> > Please store the `c_nonce` as it is needed in later steps and will be referenced as `C_NONCE`.

Reviewer: C_NONCE is never used in the rest of this document.

  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer $ACCESS_TOKEN' \
  -H 'SWIYU-API-Version: 2' \
  -d '{
    "credential_configuration_id": "university_example_sd_jwt",
    "proofs": {
        "jwt": [
          "eyJ0eXAiOiJvcGVuaWQ0dmNpLXByb29mK2p3dCIsImFsZyI6IkVTMjU2IiwiandrIjp7Imt0eSI6IkVDIiwidXNlIjoic2lnIiwiY3J2IjoiUC0yNTYiLCJraWQiOiJUZXN0LUtleSIsIngiOiJtMDhWTkNBY1FSVGNBRVFrdlk5SUwzMHpXeG42Zk5UM2NTNVBvc0JqdGZ3IiwieSI6IjlkZUZKSVR0dWZkNGlwZXdOTEduMWVYRUpaSlBCR1AxTGtWV3dQZGloRnciLCJpYXQiOjE3NTQ2NDMzNjh9fQ.eyJhdWQiOiJodHRwOi8vbG9jYWxob3N0OjgwODAvb2lkNHZjaSIsIm5vbmNlIjoiY2MxYmI4N2MtOTA3MC00OWE4LWFmMTYtNTA3MTkwZjg0ZmUyIiwiaWF0IjoxNzU0NjQzMzY4fQ.MFuxwr_TnEU6xqgaXkXaS0_6KRWSVd8LDPmkSUkMMrJlGHjL_m56B6TYiijZ_1HmpuS89hBSGT9xEsUbh-ShjA"
        ]
    }
}'
```

And the wallet receives a transaction ID in the response:

```json
{
    "transaction_id": "3c2e23f5-6aea-4478-b4a0-304fdc1b4933",
    "interval": 5
}
```

> [!NOTE]
> Please store the `transaction_id` as it is needed in later steps and will be referenced as `$TRANSACTION_ID`.

### Business Issuer checks the offer updates status and sets the offer data (if desired)

Actor: Business Issuer

> The business issuer can then either update the status of the offer (if the offer data is already set or not)

Reviewer: ```suggestion
The business issuer can then decide to update the status of the offer (if the offer data is already set) or not.
```
